{% extends 'base.html' %}

{% load bootstrap %}
{% load render_table from django_tables2 %}
{% block content %}

    <h1>SendSketch Results for Sample {{ base_project.title }}</h1>
    <br>
    {% if base_project.sendsketch_status == 'Processing' %}
        <pre>Processing...</pre>
        <i class="fa fa-circle-o-notch fa-spin" style="font-size:36px"></i>
    {% elif base_project.sendsketch_status == 'Complete' %}
        <table id="sendsketch_results" class="table table-hover table-sm table-responsive table-bordered">
        <thead>
            <tr>
                <th><button class="btn btn-secondary btn-sm btn-block" data-toggle="tooltip" title="Rank according to ANI.">Rank</button></th>
                <th><button class="btn btn-secondary btn-sm btn-block" data-toggle="tooltip" title="Weighted Kmer IDentity, which is the kmer identity compensating for differences in size.  So, comparing human chr1 to the full human genome would yield 100% WKID but approximately 10% KID.">WKID</button></th>
                <th><button class="btn btn-secondary btn-sm btn-block" data-toggle="tooltip" title="Kmer IDentity, equal to matches/length; this is the fraction of shared kmers.">KID</button></th>
                <th><button class="btn btn-secondary btn-sm btn-block" data-toggle="tooltip" title="Average Nucleotide Identity, derived from WKID and kmer length.">ANI</button></th>
                <th><button class="btn btn-secondary btn-sm btn-block" data-toggle="tooltip" title="Genome completeness (percent of the reference represented in the query).  Derived from WKID and KID.">Complt</button></th>
                <th><button class="btn btn-secondary btn-sm btn-block" data-toggle="tooltip" title="Contamination (percent of the query that does not match this reference, but matches some other reference).">Contam</button></th>
                <th><button class="btn btn-secondary btn-sm btn-block" data-toggle="tooltip" title="The number of shared kmers between query and ref.">Matches</button></th>
                <th><button class="btn btn-secondary btn-sm btn-block" data-toggle="tooltip" title="The number of shared kmers between query and ref, and no other ref.">Unique</button></th>
                <th><button class="btn btn-secondary btn-sm btn-block" data-toggle="tooltip" title="Number of kmers that did not hit any reference sequence.  Though constant for a query, it will be reported differently for different references based on the relative size of the reference and query (if the reference is bigger than the query, it will report all of them).">noHit</button></th>
                <th><button class="btn btn-secondary btn-sm btn-block" data-toggle="tooltip" title="NCBI taxonomic id, when available.">TaxID</button></th>
                <th><button class="btn btn-secondary btn-sm btn-block" data-toggle="tooltip" title="Estimate of genomic size (number of unique kmers in the genome).  This is based on the smallest hash value in the list.  This is affected by blacklists or whitelists, and by using an assembly versus raw reads.">gSize</button></th>
                <th><button class="btn btn-secondary btn-sm btn-block" data-toggle="tooltip" title="Number of sequences used in the sketch.">gSeqs</button></th>
                <th><button class="btn btn-secondary btn-sm btn-block" data-toggle="tooltip" title="NCBI's name for that taxID.  If there is no taxID, the sequence name will be used.">taxName</button></th>
            </tr>
        </thead>

        <tbody>
            {% for item in project %}
                <tr align="center">
                    <td>{{ item.rank }}</td>
                    <td>{{ item.wkid }}</td>
                    <td>{{ item.kid }}</td>
                    {% if item.ani|slice:':2'|add:"0" > 95 %}
                        <td class="table-success">{{ item.ani }}</td>
                    {% elif item.ani|slice:':2'|add:"0" > 90 %}
                        <td class="table-primary">{{ item.ani }}</td>
                    {% else %}
                        <td class="table-warning">{{ item.ani }}</td>
                    {% endif %}
                    <td>{{ item.complt }}</td>
                    {% if item.contam|slice:':2'|add:"0" > 10 %}
                        <td class="table-danger">{{ item.contam }}</td>
                    {% else %}
                        <td class="table-success">{{ item.contam }}</td>
                    {% endif %}
                    <td>{{ item.matches }}</td>
                    <td>{{ item.unique }}</td>
                    <td>{{ item.nohit }}</td>
                    <td><a href="https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id={{ item.taxid }}">{{ item.taxid }}</a></td>
                    <td>{{ item.gsize }}</td>
                    <td>{{ item.gseqs }}</td>
                    <td>{{ item.taxname }}</td>
                </tr>
            {% endfor %}
        </tbody>
        </table>
        <script src="https://cdn.datatables.net/1.10.4/js/jquery.dataTables.min.js"></script>
        <link rel="stylesheet" href="https://cdn.datatables.net/1.10.4/css/jquery.dataTables.min.css">

        <script src="https://code.jquery.com/jquery-1.11.1.min.js"></script>
        <script type="text/javascript">
            $(document).ready(function () {
                $('#sendsketch_results').dataTable();
            });
        </script>
        <pre>*Hover your cursor over the column names to display a full description.</pre>
        {% else %}
        <pre>Not available</pre>
{% endif %}

{% endblock %}
